classes/config_settings.php

```diff
@@ -0,0 +1,259 @@
+     */
+    public function __construct(int $quizid, \stdClass $settings = null) {
+        global $DB;
+
+        $this->quizid = $quizid;
+
+        // Try and retrieve existing settings, otherwise assign null.
+        $existingsettings = $DB->get_record(self::SETTINGS_TABLE, ['quizid' => $quizid], '*');
+        if (empty($existingsettings)) {
+            $existingsettings = new \stdClass();
+        } else {
+            $this->existingsettingsid = $existingsettings->id;
+        }
+
+        // If new settings are empty, set existing settings, otherwise merge existing settings with new settings.
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
+        } else {
+            $this->settings = $this->merge_new_and_existing_settings($this->strip_prefix($settings), $existingsettings);
+        }
+
+        // Associate settings with quiz.
+        $this->settings->quizid = $quizid;
+
+        // TODO: Process sebconfigtemplate into templateid.
+        $this->settings->templateid = 0;
+        unset($this->settings->sebconfigtemplate);
+
+        // Check to make sure all required settings are set up.
+        foreach ($this->get_required_settings_keys() as $name) {
+            if (!array_key_exists($name, $this->settings)) {
+                throw new \moodle_exception('missingrequiredsettings', 'quizaccess_seb', '', '', "Missing setting: $name");
+            }
+        }
+    }
+
+    /**
+     * Static constructor to create settings from existing quiz.
+     *
+     * @param string $quizid ID of quiz that the settings belong to.
+     * @return config_settings This instance.
+     *
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public static function with_quizid(string $quizid) {
+        return new self($quizid);
+    }
+
+    /**
+     * Static constructor to create settings from form data.
+     *
+     * @param \stdClass $quiz New quiz settings. May be provided by quiz settings form. Must include 'id'.
+     * @return config_settings This instance.
+     *
+     * @throws \dml_exception
+     * @throws \invalid_parameter_exception
+     * @throws \moodle_exception
+     */
+    public static function with_form_data(\stdClass $quiz) {
+        $settings = new \stdClass;
+        if (empty($quiz->id)) {
+            throw new \invalid_parameter_exception('Cannot construct config_settings without $quiz->id');
+        }
+
+        foreach ($quiz as $key => $quizsetting) {
+            if (strpos($key, 'seb_') === 0) {
+                $settings->$key = $quizsetting;
+            }
+        }
+
+        return new self($quiz->id, $settings);
+    }
+
+    /**
+     * Check if the settings exist.
+     *
+     * @return bool Whether any new or existing settings were found for quiz.
+     */
+    public function exists() {
+        // Cast settings to array as empty stdClass will return true.
+        return !empty((array) $this->settings);
```

Comment:
What would happen if I create a class with bunch of random settings? Do we need to hardcore settings we support (e.g. settings we  have fields for in DB)?

Reply:
Ah I see you have get_required_settings_keys! :+1: